### PR TITLE
model access control documentation

### PR DESCRIPTION
**Pull request by rbhavna**
### Description
model access control documentation
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/model_access_control.md [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+
+We have implemented backend role-based access in which members of the same back-end roles will be able to access each other’s resources. This is similar to alerting and anomaly detection plugin [backend role based access](https://opensearch.org/docs/latest/observing-your-data/alerting/security/#advanced-limit-access-by-backend-role). Additionally we also let users define an ml resource as public or private. This advanced security access is available only on a cluster with [Security plugin](https://opensearch.org/docs/latest/security/index/).
+
+Note: 
+- Model access control is experimental feature. If you see any bug or have any suggestion, feel free to cut Github issue.
```
Comment (dhrubo-os) [inline docs/model_access_control.md:13]: May be add a link for the github issue?

File: docs/model_access_control.md [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+
+
+
+## Setup
+
+For Backend Role based security, it is not enabled by default. To enable/disable it, customers can toggle elasticsearch cluster setting shown below. If this setting is not enabled, a resource can be accessed by anyone.
```
Comment (dhrubo-os) [inline docs/model_access_control.md:19]: Just curious, to create a model group, do we need to enable this setting?

File: docs/model_access_control.md [hunk 3]
```diff
@@ -0,0 +1,652 @@
+# Model Access Control
+
+In the present setting, on a security enabled OpenSearch cluster, we can limit users to certain actions by mapping them to relevant permissions and permission roles. For example, we may want some users to only search and view ml models while allow others to register and deploy them.
```
Comment (ylwu-amzn) [inline docs/model_access_control.md:3]: "we can limit users..." -> "admin can limit user..."

"we may want some users... " -> "admin can grant some user access to .."

File: docs/model_access_control.md [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+
+The access control in ml-commons is implemented at the _model group level_ so that all versions of the same model share the same access mode as that of their model group. That is, if user has access to one version, then he/she should have access to all versions of the model group.
+
+When creating a model group, user should also specify the access mode to it. Depending on the access mode, other users will be able to access the versions of a model group or create new versions to it.
+
+If a user tries to register/deploy/undeploy/predict/delete/search/get model versions, first we check if the user has access to the model. Otherwise, we throw exception saying user does not have access to the model specified.
+
+A user is considered a model _owner_ when he/she creates a new model group and register its first version. When a model owner creates a model group, the owner can specify one of the following _access modes_ for this model group:
```
Comment (dhrubo-os) [inline docs/model_access_control.md:52]: `register its first version` : how about register the first version of the model?

File: docs/model_access_control.md [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+- `private`: Only the model owner or an admin user can access this model group.
+- `restricted`: The owner, an admin user, or any user who shares one of the model group's backend roles can access any model in this model group. When creating a `restricted` model group, the owner must attach one or more of the owner's backend roles to the model. For a user to be able to access a model group, one of the user's backend roles must match one of the model group's backend roles. This grants the user permissions associated with the user role to which that backend role is mapped. For example, if the backend role is mapped to the `ml_full_access` user role, users with this role can use the Register, Deploy, Predict, Delete, Search and Get Model APIs on any model in this model group.
+
+An admin can access all model groups in the cluster regardless of their access mode.
+
+
+
+## Registering a model group:
+
+User should use the `_register` endpoint to register a model group. User can register a model group with a `public`, `private`, or `restricted` access mode.
```
Comment (dhrubo-os) [inline docs/model_access_control.md:64]: `_register` api endpoint

File: docs/model_access_control.md [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+| `name`                  | String    | The model group name. Required.                                                                                                                                                                                                                           |
+| `description`           | String    | The model group description. Optional.                                                                                                                                                                                                                    |
+| `model_access_mode`     | String    | The access mode for this model. Valid values are `public`, `private`, and `restricted`. When this parameter is set to `restricted`, user must specify either `backend_roles` or `add_all_backend_roles`, but not both. Optional. Default is `restricted`. |
+| `backend_roles`         | Array     | A list of the model owner's backend roles to add to the model. Can be specified only if the `model_access_mode` is `restricted`. Cannot be specified at the same time as `add_all_backend_roles`. Optional.                                               |
```
Comment (dhrubo-os) [inline docs/model_access_control.md:81]: Can be specified only if the `model_access_mode` is `restricted`. --> Required only if the `model_access_mode` is `restricted` ?

File: docs/model_access_control.md [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+
+A user is considered a model _owner_ when he/she creates a new model group and register its first version. When a model owner creates a model group, the owner can specify one of the following _access modes_ for this model group:
+
+- `public`: All users who have access to the cluster can access this model group.
+- `private`: Only the model owner or an admin user can access this model group.
+- `restricted`: The owner, an admin user, or any user who shares one of the model group's backend roles can access any model in this model group. When creating a `restricted` model group, the owner must attach one or more of the owner's backend roles to the model. For a user to be able to access a model group, one of the user's backend roles must match one of the model group's backend roles. This grants the user permissions associated with the user role to which that backend role is mapped. For example, if the backend role is mapped to the `ml_full_access` user role, users with this role can use the Register, Deploy, Predict, Delete, Search and Get Model APIs on any model in this model group.
```
Comment (ylwu-amzn) [inline docs/model_access_control.md:56]: "This grants the user permissions associated with the user role to which that backend role is mapped. For example, if the backend role is mapped to the `ml_full_access` user role, users with this role can use the Register, Deploy, Predict, Delete, Search and Get Model APIs on any model in this model group."

Seems not related with `restricted`. Remove these or move to other part?

File: docs/model_access_control.md [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+```
+POST /_plugins/_ml/model_groups/_register
+{
+    "name": "model_group_test",
+    "description": "This is an example description",
+    "model_access_mode": "restricted",
+    "backend_roles" : ["IT"]
+}
+```
+
```
Comment (dhrubo-os) [inline docs/model_access_control.md:133]: May be add a comment here: "Or we can specify `add_all_backend_roles` instead of `backend_roles`?

File: docs/model_access_control.md (hunk 4 above)
Comment (ylwu-amzn) [inline docs/model_access_control.md:52]: > A user is considered a model _owner_ when he/she creates a new model group and register its first version.

This sentence says user need to create new group and register first version. What if user register model version directly to a model group? Will that user be model owner of the model version ?

File: docs/model_access_control.md [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+## Updating a model group
+
+Updating a model group request is very similar to register model group request. Additionally, user needs to specify the `model_group_id` field in the request to which he/she needs to make an update.
+
+
+### Path and HTTP method
+
+```
+PUT /_plugins/_ml/model_groups/<model_group_id>_update
```
Comment (dhrubo-os) [inline docs/model_access_control.md:184]: UT /_plugins/_ml/model_groups/<model_group_id>/_update

File: docs/model_access_control.md (hunk 4 above)
Comment (ylwu-amzn) [reply docs/model_access_control.md:52]: If I understand correctly model owner is something related with model group. If a user register a model group , the user will be the owner of the model group.

Let's user "model group" and "model version" to avoid confusion?

File: docs/model_access_control.md [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+    "model_format": "TORCH_SCRIPT",
+    "model_group_id": "FTNlQ4gBYW0Qyy5ZoxfR",
+    "model_content_hash_value": "9376c2ebd7c83f99ec2526323786c348d2382e6d86576f750c89ea544d6bbb14",
+    "model_config": {
+        "model_type": "bert",
+        "embedding_dimension": 384,
+        "framework_type": "sentence_transformers",
+       "all_config": "{\"_name_or_path\":\"nreimers/MiniLM-L6-H384-uncased\",\"architectures\":[\"BertModel\"],\"attention_probs_dropout_prob\":0.1,\"gradient_checkpointing\":false,\"hidden_act\":\"gelu\",\"hidden_dropout_prob\":0.1,\"hidden_size\":384,\"initializer_range\":0.02,\"intermediate_size\":1536,\"layer_norm_eps\":1e-12,\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6,\"pad_token_id\":0,\"position_embedding_type\":\"absolute\",\"transformers_version\":\"4.8.2\",\"type_vocab_size\":2,\"use_cache\":true,\"vocab_size\":30522}"
+    },
+    "url": "https://github.com/ylwu-amzn/ml-commons/blob/2.x_custom_m_helper/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/all-MiniLM-L6-v2_torchscript_sentence-transformer.zip?raw=true"
```
[the remark this answers is not in the document]
Comment (rbhavna) [reply docs/model_access_control.md:612]: Thanks for the catch. I used the old request accidentlly

File: docs/model_access_control.md [hunk 11]
```diff
@@ -0,0 +1,652 @@
...
+            ],
+            "name": "user1",
+            "user_requested_tenant": null
+          },
+          "created_time": 1684362035938,
+          "latest_version": 2,
+          "last_updated_time": 1684362571300,
+          "name": "model_group_test",
+          "description": "This is an example description",
+          "tags": {
```
Comment (ylwu-amzn) [inline docs/model_access_control.md:525]: Do we have tags in 2.8?

File: docs/model_access_control.md [hunk 12]
```diff
@@ -0,0 +1,652 @@
...
+
+
+#### Example Request: Registering a model version via URL
+
+```
+POST /_plugins/_ml/models/_register
+
+{
+    "name": "all-MiniLM-L6-v2",
+    "version": "1.0.0",
```
Comment (ylwu-amzn) [inline docs/model_access_control.md:601]: version is not needed here, right? You can explain more about version part

File: docs/model_access_control.md (hunk 11 above)
Comment (rbhavna) [reply docs/model_access_control.md:525]: Accidentally used the old response. Will change it

File: docs/model_access_control.md (hunk 12 above)
Comment (rbhavna) [reply docs/model_access_control.md:601]: So the auto incrementing of the version?

File: docs/model_access_control.md [hunk 13]
```diff
@@ -0,0 +1,652 @@
...
+### Path and HTTP method
+
+```
+PUT /_plugins/_ml/model_groups/<model_group_id>_update
+```
+
+* A user can make updates to a model group to which he/she has access which is determined by the access mode of the model group.
+* In a model_access_control enabled cluster, the model group owner or an admin user can update all fields. Whereas, any other user who shares one or more backend roles with the model group can update the `name` and `description` fields only.
+* In a security/model_access_control disabled cluster, any user can make updates to any model group but cannot specify any of the access fields.
+* Consider a model group created by User1 with IT backend_role set to it. Since user1 is the owner, he/she will be able to update all the fields of the model group.
```
Comment (ylwu-amzn) [inline docs/model_access_control.md:190]: Seems line 188 already covers line 190, 191
Line 187 covers line 192?

Line 190, 191, 192 are describing an example. Remove these lines or move these lines out of bulleted list?

File: docs/model_access_control.md [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+* If User2 tries to send update request to the same model group, only name, and description will be allowed to be updated but not other fields because user2 has access to the model group but is not the owner/admin.
+* If User3 tries to send update request, exception will be thrown because user3 has no matching backend_roles of the model group and therefore has no access to it at all.
+
+Similar to registering, there are other checks for updating a model group.
```
Comment (ylwu-amzn) [inline docs/model_access_control.md:194]: These exceptions seems not the most important things for most users who are using this feature in correct way. How about we add a new section for all checkings and exceptions like register and update model group? So we can keep the doc clean and tight for most happy cases. Otherwise, user have to read through a long document to start use, they may be lost in too much details.

File: docs/model_access_control.md [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+
+For example:
+* For model versions registered in a model group created by User1 with IT backend_role, only user1, user2 and admin can perform the above actions. If user3 and user4 try to send a request, an exception will be thrown: `You don't have permissions to perform this operation on this model.`
+* For model versions registered in a private model group of User1, only User1 or admin will be able to perform the above actions on its versions. For any other user, above exception will be thrown.
+* For a public model group of User1, any user can perform the said actions its versions.
+
+
+
+
+## Search model versions
```
Comment (ylwu-amzn) [inline docs/model_access_control.md:645]: Add some example search request? Like how to search all model versions in a model group with model group id

File: docs/model_access_control.md [hunk 16]
```diff
@@ -0,0 +1,652 @@
...
+#Response
+{
+    "task_id": "FjNlQ4gBYW0Qyy5Zxhdh",
+    "status": "CREATED"
+}
+```
+
+
+
+## Deploy/Undeploy/Predict/Delete/Get model version
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply docs/model_access_control.md:626]: You need to fix this doc https://github.com/opensearch-project/ml-commons/blob/2.x/docs/model_serving_framework/text_embedding_model_examples.md, this doc doesn't have model group id part

[the remark this answers is not in the document]
Comment (rbhavna) [reply docs/model_access_control.md:626]: Yeah so there is basically no change to API requests here. Maybe we can add a link

File: docs/model_access_control.md (hunk 2 above)
Comment (rbhavna) [reply docs/model_access_control.md:19]: not necessary for creating model group. This setting is only to enable access control

File: docs/model_access_control.md (hunk 14 above)
Comment (rbhavna) [reply docs/model_access_control.md:194]: Actually I have removed the exception cases from this doc. As you mentioned, this might cause unnecessary confusion to the user. Just made sure that user knows which field should be specified in what case

**Review by dhrubo-os — APPROVED** [review]

**Review by Zhangxunmt — APPROVED** [review]

File: docs/model_access_control.md (hunk 2 above)
Comment (dhrubo-os) [reply docs/model_access_control.md:19]: hmm, I just tried to create a model group:

```
POST /_plugins/_ml/model_groups/_register
{
    "name": "test_model_group_public",
    "description": "This is a public model group",
    "model_access_mode": "public"
}
```

And received error :

```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "Cluster security plugin not enabled or model access control no enabled, can't pass access control data in request body"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "Cluster security plugin not enabled or model access control no enabled, can't pass access control data in request body"
  },
  "status": 400
}
```

But after enabling the `plugins.ml_commons.model_access_control_enabled` then I was able to create the model group:

```
{
  "model_group_id": "rJTkoYgBU4BPBnjEiV4f",
  "status": "CREATED"
}
```

Comment (rbhavna) [reply docs/model_access_control.md:19]: So here, in this case, all model groups are public by default. We do not want user to specify the model_access_mode when the setting/security is disabled. Try this without the setting enabled. It would work
